hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterServiceImpl.java

```diff
@@ -192,125 +166,173 @@ public void init(NodeEngine nodeEngine, Properties properties) {
+            logger.warning("Cannot send explicit suspicion trigger for " + endpointMembersViewMetadata + " to itself.");
             return;
         }
 
+        int memberListVersion = membershipManager.getMemberListVersion();
+        Operation op = new TriggerExplicitSuspicionOp(memberListVersion, endpointMembersViewMetadata);
+        OperationService operationService = nodeEngine.getOperationService();
+        operationService.send(op, triggerTo);
+    }
+
+    public MembersView handleMastershipClaim(Address candidateAddress, String candidateUuid) {
+        // verify candidateAddress is not me DONE
+        // verify I am not master DONE
+        // verify candidateAddress is a valid member with its uuid
+        // verify I suspect everyone before the candidateAddress
+        // verify candidateAddress is not suspected.
+
+        checkNotNull(candidateAddress);
+        checkNotNull(candidateUuid);
+        checkFalse(getThisAddress().equals(candidateAddress), "cannot accept my own mastership claim!");
+
         lock.lock();
         try {
-            if (deadAddress.equals(node.getMasterAddress())) {
-                assignNewMaster();
-            }
-            if (node.isMaster()) {
-                clusterJoinManager.removeJoin(deadAddress);
+            checkTrue(isJoined(), candidateAddress + " claims mastership but this node is not joined!");
+            checkFalse(isMaster(),
+                    candidateAddress + " claims mastership but this node is master!");
+            checkFalse(candidateAddress.equals(getMasterAddress()),
+                    candidateAddress + " claims mastership but it is already the known master!");
+
+            MemberImpl masterCandidate = membershipManager.getMember(candidateAddress, candidateUuid);
+            checkTrue(masterCandidate != null ,
+                    candidateAddress + " claims mastership but it is not a member!");
+
+            MemberMap memberMap = membershipManager.getMemberMap();
+            if (!shouldAcceptMastership(memberMap, masterCandidate)) {
+                String message = "Cannot accept mastership claim of " + candidateAddress
+                        + " at the moment. There are more suitable master candidates in the member list.";
+                logger.fine(message);
+                throw new RetryableHazelcastException(message);
             }
-            Connection conn = node.connectionManager.getConnection(deadAddress);
-            if (destroyConnection && conn != null) {
-                conn.close(reason, null);
-            }
-            MemberImpl deadMember = getMember(deadAddress);
-            if (deadMember != null) {
-                removeMember(deadMember);
-                logger.info(membersString());
+
+            if (!membershipManager.clearMemberSuspicion(candidateAddress, "Mastership claim")) {
+                throw new IllegalStateException("Cannot accept mastership claim of " + candidateAddress + ". "
```

Comment:
Members: [A, B, C, D]
B suspects A, C suspects [A, B], D suspects [A, B]. B and C start the mastership claim process. D first receives mastership claim of C and accepts it since it is a valid one. Then, D receives a mastership claim from B. At that point, it will reject it because it is too late to accept mastership claim of B since D has already accepted a younger node as its master: which is C.